ENH: add geomspace function

endolith opened the pull request:
Like logspace, but specifying start and stop directly, without having to
specify the base.

Fixes #7255

njsmith commented:
Please add link to mailing list discussion for future reference.

endolith commented:
Added tests, please review

madphysicist commented on numpy/core/function_base.py:
```diff
@@ -206,3 +206,76 @@ def logspace(start, stop, num=50, endpoint=True, base=10.0, dtype=None):
+    --------
+    logspace : Similar to geomspace, but with endpoints specified using log
+               and base.
+    linspace : Similar to geomspace, but with the samples uniformly distributed
```
I think "Similar to `geomspace` but with arithmetic instead of geometric sequence."
would be better.

endolith replied:
ok but this was copied from the other functions, should those change too?

madphysicist replied:
I would say let someone else weigh in, preferably a core dev, then make all of them consistent one way or the other.

endolith replied:
changed only in geomspace

madphysicist commented on numpy/core/function_base.py:
```diff
@@ -206,3 +206,76 @@ def logspace(start, stop, num=50, endpoint=True, base=10.0, dtype=None):
+    linspace : Similar to geomspace, but with the samples uniformly distributed
+               in linear space, instead of log space.
+    arange : Similar to linspace, with the step size specified instead of the
+             number of samples. Note that, when used with a float endpoint, the
```
Remove "note that" from the beginning of the sentence.

madphysicist replied:
In fact, I am not sure that sentence is necessary at all. This is just a see also section. The curious will follow the link to the full, detailed documentation.

endolith replied:
changed

madphysicist commented on numpy/core/function_base.py:
```diff
@@ -206,3 +206,76 @@ def logspace(start, stop, num=50, endpoint=True, base=10.0, dtype=None):
+    >>> import matplotlib.pyplot as plt
+    >>> N = 10
+    >>> x1 = np.geomspace(1, 10, N, endpoint=True)
+    >>> x2 = np.geomspace(1, 10, N, endpoint=False)
```
Would you consider plotting `x` as `0:N` and `0:N-1` (just the index), `y1` and `y2` as the results of `geomspace`? I think that the vertical spacing might be more illustrative in that case.

endolith replied:
probably a good idea

madphysicist replied:
Also, it might be good to display on a semilog plot.

endolith replied:
I'm not sure I this is an improvement or not:

![figure_1](https://cloud.githubusercontent.com/assets/58611/13589397/a6095a34-e4a5-11e5-9bb3-1083d3bf4563.png)

madphysicist replied:
Odd. I was expecting a straight line. For reference, something like `plt.semilogy(np.arange(n), np.geomspace(10, 100, n))`

endolith replied:
yours:

![figure_1-2](https://cloud.githubusercontent.com/assets/58611/13622182/027d8ca2-e56b-11e5-8673-ca131d6274d3.png)

another option:

![figure_1-1](https://cloud.githubusercontent.com/assets/58611/13622127/9ec5c378-e56a-11e5-9cb9-c74375e553b1.png)

madphysicist replied:
Personally, I think my option is more illustrative. I will go with whatever you choose though.

madphysicist commented on numpy/core/function_base.py:
```diff
@@ -206,3 +206,76 @@ def logspace(start, stop, num=50, endpoint=True, base=10.0, dtype=None):
+    """
+    if start == 0:
+        raise ValueError('Geometric sequence starting point must be nonzero')
+    return logspace(_nx.log10(start), _nx.log10(stop), num=num,
```
Just out of curiosity, what motivated you to pick base 10 over say base e? I am not disagreeing, just genuinely curious.

endolith replied:
1. It's the default in logspace
2. It makes common numbers work out to exact integers.  The existing `TestLogspace` tests for exact equality, for instance:

```
y = logspace(0, 6, num=7)
assert_array_equal(y, [1, 10, 100, 1e3, 1e4, 1e5, 1e6])
```

which does not work for other bases:

```
logspace(log10(1), log10(1e3), num=4)
Out[125]: array([    1.,    10.,   100.,  1000.])

logspace(log10(1), log10(1e3), num=4) == array([1, 10, 100, 1000])
Out[126]: array([ True,  True,  True,  True], dtype=bool)

logspace(log2(1), log2(1e3), num=4, base=2)
Out[127]: array([    1.,    10.,   100.,  1000.])

logspace(log2(1), log2(1e3), num=4, base=2) == array([1, 10, 100, 1000])
Out[128]: array([ True, False, False,  True], dtype=bool)
```

madphysicist replied:
Good choice. Thanks for the detailed explanation.

madphysicist commented on numpy/core/tests/test_function_base.py:
```diff
@@ -29,6 +29,30 @@ def test_dtype(self):
+        assert_equal(y.dtype, dtype('float64'))
+        y = geomspace(1, 1e6, dtype='int32')
+        assert_equal(y.dtype, dtype('int32'))
+
```
A couple of additional tests I would like to see:

  - `start` > `stop`
  - complex inputs
  - negative `start`, `stop`, both
  - Negative stop should raise too

madphysicist replied:
You may also want to add basically the same tests as `linspace`/`logspace` do, possibly including the ones for subclasses, etc. While currently the tests of `logspace` cover most of the functionality of your function, there is no guarantee that it will be implemented in terms of `logspace` in the future. For example, the `exp(linspace(...))` implementation given by @rkern [here](https://github.com/numpy/numpy/issues/7255#issuecomment-184586446). `numpy.percentile` used to be implemented in terms of `sort`. Then it got reimplemented in terms of `numpy.partition`. That is the sort of thing tests are really good for.

madphysicist replied:
I am a bit of a fan of overkill testing. For every time a test has saved me from publicly showing my stupidity, there has always been that one test that I forgot to write.

endolith replied:
How should complex or negative inputs behave?

`geomspace(1+0j, 0-1j, 4)` produces

[1+0j, √3/2-0.5j, 0.5-√3/2j,   0-1j]  (each √3/2-0.5j times the previous)

but

[1+0j,  0+1j, -1+0j,  0-1j]  (each 1j times the previous)

would be equally valid, I think (or any other logarithmic spiral that connects the 2 points?)

madphysicist replied:
At the moment, you are doing whatever `logspace` is doing, and that is fine. I do like the idea of supporting negative numbers though. Either way, you should define and document the behavior, then write a test for it.

madphysicist replied:
You may want to ask for the opinion of someone who might use such a feature. I can see myself using this function, but only with real inputs, so I would be fine with complex and negative inputs raising an error, returning nan or connecting the endpoints via some meaningful curve. The last option is probably optimal. The point of documenting and writing tests would be that if you change the internals, the public API remains the same and there are tests to make sure that happens.

madphysicist commented on numpy/core/function_base.py:
```diff
@@ -206,3 +206,76 @@ def logspace(start, stop, num=50, endpoint=True, base=10.0, dtype=None):
+    >>> plt.show()
+
+    """
+    if start == 0:
```
Zero `stop` should raise too. I am guessing that swapping the limits reverses the samples. If that's true, stop should be checked explicitly.

endolith replied:
Good point, it does.

madphysicist commented:
I have a few comments, but overall +1 for this. This is much more intuitive than logspace.

endolith commented:
Mailing list proposals:

- Different name?
  - I think `geomspace` is great
- `np.linspace(start, stop, steps, spacing="log")`
 - I really don't like this.  Just change the function name to `logspace` if you want log space.
- Specify number of intervals rather than number of steps?
 - maybe, but would be confusing coming from existing functions
- option for `geomspace(start=1, finish=2, ratio=1.03)`
 - I agree that this is useful, or `geomspace(start=1, num=7, ratio=1.03)`. However...
- "call a function with this signature `geomrange()` instead"
 - Agreed.  Separate functions is better than a single Swiss army knife function that tries to do everything with mutually-incompatible parameters.

endolith commented:
I added support for negative start and stop, but this breaks some complex start and stop, and I'm not sure whether those should be supported anyway.  An implementation that doesn't use logspace could support both I guess.

madphysicist commented:
+1 to two functions. I pretty much agree with the whole explanatory post.

charris commented:
How about `expspace` as a name? AFAICT, what we have is basically `exp(a*linspace(...))`. As a generalization, if `f` is monotone, `f(a*linspace(...)`, so there could be a general underlying function with special cases, the only thing really needed is the ability to invert `f` to get the endpoints.

endolith commented:
Ok I modified it to handle negative and complex arguments.  With complex log, it actually works out without any special-casing, but I special-cased negative arguments so the output doesn't have negligible imaginary parts.  I could also special-case purely-imaginary arguments so they don't have negligible real parts.

Currently:

good:

```
geomspace(1, 1000, 4)
 array([    1.,    10.,   100.,  1000.])

geomspace(-1, -1000, 4)
 array([   -1.,   -10.,  -100., -1000.])

geomspace(1, 16, 5)
 array([  1.,   2.,   4.,   8.,  16.])

geomspace(1, 1000, 4, dtype=int)
 array([   1,   10,  100, 1000])

geomspace(-1, -1000, 4, dtype=int)
 array([   -1,   -10,  -100, -1000])
```

reasonable:

```
geomspace(1, -1, 5)  # dtype derived from real arguments
 array([ nan,  nan,  nan,  nan,  nan])

geomspace(1+0j, -1, 5)  # dtype derived from complex argument
array([  1.00000000e+00 +0.00000000e+00j,
         7.07106781e-01 +7.07106781e-01j,
         6.12323400e-17 +1.00000000e+00j,
        -7.07106781e-01 +7.07106781e-01j,  -1.00000000e+00 +1.22464680e-16j])

geomspace(1, -1+0j, 5)  # dtype derived from complex argument
array([  1.00000000e+00 +0.00000000e+00j,
         7.07106781e-01 +7.07106781e-01j,
         6.12323400e-17 +1.00000000e+00j,
        -7.07106781e-01 +7.07106781e-01j,  -1.00000000e+00 +1.22464680e-16j])

geomspace(1, -1, 5, dtype=complex)  # dtype specified
array([  1.00000000e+00 +0.00000000e+00j,
         7.07106781e-01 +7.07106781e-01j,
         6.12323400e-17 +1.00000000e+00j,
        -7.07106781e-01 +7.07106781e-01j,  -1.00000000e+00 +1.22464680e-16j])

geomspace(1j, 1000j, 4)
Out[250]: 
array([  6.12323400e-17 +1.00000000e+00j,
         6.12323400e-16 +1.00000000e+01j,
         6.12323400e-15 +1.00000000e+02j,   6.12323400e-14 +1.00000000e+03j])

geomspace(-1j, -1000j, 4)
Out[251]: 
array([ -6.12323400e-17 -1.00000000e+00j,
        -6.12323400e-16 -1.00000000e+01j,
        -6.12323400e-15 -1.00000000e+02j,  -6.12323400e-14 -1.00000000e+03j])        
```

bad:

```
geomspace(1, 16, 5, dtype=int)
 array([ 1,  2,  4,  7, 16])
                     ^
```

probably should check if dtype is int and use around() first?

madphysicist commented:
With regards to your "bad" example, you are not obligated to return an array of the same type as the input, and I would say that you definitely should not. Many `numpy` functions (e.g. `percentile` comes to mind) will return an array of a reasonable type able to hold the result. In your case, returning a `float64` real array for integer inputs is probably not an unreasonable solution. Returning an integer array is unreasonable in almost any case for this function for the reason shown above.

endolith commented:
the bad example explicitly states dtype=int, though.  if dtype is not specified, it's either float or complex output

madphysicist commented:
Looks like you will have to do your own rounding then.

seberg commented:
You could also just kill the dtype argument ;). This is unsafe casting, ufuncs will nowadays not allow you to use dtype=int, i.e. for log it will tell you that there is no loop for this type and it will tell you it is not possible for float + float = int, unless you also specify casting=unsafe (though it will use float internally then).

madphysicist commented:
Killing `dtype` is probably better. Here is another bad case: Inputs are complex, output dtype is real. `around` won't help with that much.

njsmith commented:
Output dtype argument is very useful for dtype=float32. Making dtype=int
raise an error is totally reasonable, though.
On Mar 7, 2016 9:07 AM, "Mad Physicist" <notifications@github.com> wrote:

> Killing dtype is probably better. Here is another bad case: Inputs are
> complex, output dtype is real. around won't help with that much.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/7268#issuecomment-193348778>.
>

endolith commented:
I don't understand why `dtype=int` should raise an error.  If someone wants a geometric integer progression, why can't they have it?  Why shouldn't `geomspace(2, 54, 4, dtype=int)` output `array(2, 6, 18, 54])`?

I just need to add a line like:

```
if np.issubdtype(dtype, np.integer):
    result = _nx.around(result)
```

madphysicist commented:
Because someone can also ask for `geomspace(2, 7, 4, dtype=int)` or `geomspace(2, 7+3j, 4, dtype=int)` and guessing how they want that is unpythonic. Let the user call `astype` or `around` for themselves as they see fit.

njsmith commented:
If someone wants an integer progression of course they can have it -- the question is whether `dtype=int` should magically do it, or whether they should just call `around` themselves if that's what they want :-). I don't think there are any other places in numpy where we give ints a special-case casting mechanism (which is basically what this rounding proposal is), so it's a bit surprising and I doubt people will realize the feature is there...

njsmith commented:
(To be clear I'm not dead-set against the `dtype=int` thing, I'm just not 100% convinced and leaning towards erring on the side of conservatism.)

endolith commented:
I modified it so that purely-imaginary sequences are handled better:

```
geomspace(1j, 1000j, 4)
Out[376]: array([ 0.   +1.j,  0.  +10.j,  0. +100.j,  0.+1000.j])

geomspace(-1j, -1000j, 4)
Out[377]: array([ 0.   -1.j,  0.  -10.j,  0. -100.j,  0.-1000.j])
```

I didn't add `around()`.

endolith commented on numpy/core/function_base.py:
```diff
@@ -87,8 +87,8 @@ def linspace(start, stop, num=50, endpoint=True, retstep=False, dtype=None):
-    start = start * 1.
-    stop = stop * 1.
+    start = start * 1.0
+    stop = stop * 1.0
```
just clarifying that this is a float

endolith commented on numpy/core/function_base.py:
```diff
@@ -206,3 +206,105 @@ def logspace(start, stop, num=50, endpoint=True, base=10.0, dtype=None):
+    result = out_sign * logspace(log_start, log_stop, num=num,
+                                 endpoint=endpoint, base=10.0, dtype=dtype)
+
+    return result.astype(dtype)
```
is this needed?  could probably just return on the above line, since dtype will be handled by logspace and out_sign?

endolith commented:
Added a bunch more tests, changed the type promotion according to http://stackoverflow.com/q/2875024/125507 so the PhysicalQuantities test passes.

homu commented:
:umbrella: The latest upstream changes (presumably #7328) made this pull request unmergeable. Please resolve the merge conflicts.

endolith commented:
I still need to make an example and squash some commits

endolith commented:
Ok I changed the example and squashed some commits, please check it

charris commented:
It would be good if the folks involved in this would take it through to completion.

endolith committed:
ENH: Add geomspace function

Like logspace, but specifying start and stop directly, without having to
specify the base.

Purely imaginary or purely negative real sequences are converted to
positive real, computed, and converted back, so there are no negligible
real or imaginary parts.  Instead of

array([  6.12323400e-17 +1.00000000e+00j,
         6.12323400e-16 +1.00000000e+01j,
         6.12323400e-15 +1.00000000e+02j,
         6.12323400e-14 +1.00000000e+03j])

it outputs array([ 0.   +1.j,  0.  +10.j,  0. +100.j,  0.+1000.j])

If dtype is complex it does the math in complex so it can leave
the real line and follow a spiral.

TST: Added tests for geomspace and more tests for logspace, making
PhysicalQuantities tests work for all types of functions

PEP8: __all__ after imports, line wrapping

endolith commented:
Anything else I need to do?  I added some more examples that will hopefully pass.

charris commented:
Might squash the commits. Two commits, implementation tests, or just one would do. I'm not a big fan of the name but no one else seems bothered by it.

endolith commented:
@charris Ok I squashed the commits.

@madphysicist 
> Because someone can also ask for geomspace(2, 7, 4, dtype=int) or geomspace(2, 7+3j, 4, dtype=int) and guessing how they want that is unpythonic.

Yes I agree now.  It would be nice if it output exact integer values when it was mathematically supposed to, though.  But that could be added in another PR.  I documented it in the Examples for now.

charris commented:
Need to add an entry in `doc/source/reference/routines.array-creation.rst` where there is already `linspace` and `logspace`, and in the 1.12.0 release notes. Otherwise, LGTM, although as @madphysicist says, the more tests the better.

Comment when ready so that github will send out a notification.

endolith committed:
DOC: Add geomspace to function list, release notes

charris commented:
Needs resquash. You can move the update commit and squash it into the initial commit.

endolith commented:
ok I did that but I can't squash the commits right now

endolith commented:
ok resquashed

charris commented:
Thanks @endolith .

endolith commented:
Also should this be tweaked so that the start and stop points are exactly equal to the function arguments? https://stackoverflow.com/q/44136851/125507

ashwinvis commented:
@endolith Also this issue #9987
